scripts/check_copyright_notice.py

```diff
@@ -99,10 +126,149 @@
+
+
+def _validate_years(
+    file: Path,
+    allowed_start_years: Tuple[int, ...],
+    start_year: int,
+    end_year: int,
+    check_end_year: bool = True,
+) -> Dict:
+    """
+    Given a file, check if the header stuff is in place.
+
+    Return True if the files has the encoding header and the copyright notice,
+    optionally prefixed by the shebang. Return False otherwise.
+
+    :param file: the file to check.
+    :param allowed_start_years: list of allowed start years
+    :param start_year: year when the file was created
+    :param end_year: year when the file was last modified
+    :param check_end_year: whether to validate end year or not
+    :return: True if the file is compliant with the checks, False otherwise.
+    """
+
+    modification_date = get_modification_date(file)
+    check_info = {
+        "check": True,
+        "message": f"Start year {start_year} is not in the list of allowed years; {allowed_start_years}.",
+        "start_year": start_year,
+        "end_year": end_year,
+        "last_modification": modification_date,
+        "error_code": ErrorTypes.NO_ERROR,
+    }
+
+    # Start year is not in allowed start years list
+    if start_year not in allowed_start_years:
+        check_info["check"] = False
+        check_info[
+            "message"
+        ] = f"Start year {start_year} is not in the list of allowed years; {allowed_start_years}."
+        check_info["error_code"] = ErrorTypes.START_YEAR_NOT_ALLOWED
+        return check_info
+
+    # Specified year is 2021 but the file has been last modified in another later year (missing -202x)
+    if end_year is not None and check_end_year:
+
+        if start_year > end_year:
+            check_info["check"] = False
+            check_info["message"] = "End year should be greater then start year."
+            check_info["error_code"] = ErrorTypes.START_YEAR_GT_END_YEAR
+            return check_info
+
+        if end_year != modification_date.year:
+            check_info["check"] = False
+            check_info[
+                "message"
+            ] = f"End year does not match the last modification year. Header has: {end_year}; Last Modified: {modification_date.year}"
+            check_info["error_code"] = ErrorTypes.END_YEAR_WRONG
+            return check_info
+
+    if end_year is None and modification_date.year > start_year:
+        check_info["check"] = False
+        check_info["message"] = f"Missing later year ({start_year}-20..)"
+        check_info["error_code"] = ErrorTypes.END_YEAR_MISSING
+        return check_info
+
+    return check_info
+
 
-def check_copyright(file: Path) -> bool:
+def _check_mixed(file: Path, content: str) -> Tuple[Dict, str]:
+    check_data = _validate_years(
+        file, START_YEARS_VALORY, *get_year_data(VALORY_REGEX.search(content), True)  # type: ignore
+    )
+    check_data["fetchai_year_data"] = get_year_data(FETCHAI_REGEX.search(content), True)  # type: ignore
+
+    return check_data, MIXED
```

Comment:
On a mixed header we'll only validate the valory part since there won't be any new changed from fetchai.